File: 0494-target-sum/0494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: list[int], target: int) -> int:
        total = sum(nums)

        if abs(target) > total:
            return 0

        new_target = total + target

        if new_target % 2 != 0:
            return 0
        
        new_target = new_target // 2
        dp = [0] * (new_target + 1)
        dp[0] = 1

        for num in nums:
            for target_val in range(new_target, -1 , -1):
                if num <= target_val:
                    dp[target_val] += dp[target_val - num]
        
        return dp[-1]

File: 0494-target-sum/0494-target-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var findTargetSumWays = function(nums, target) {
    let cache = new Map();
    return findWays(0, 0);
    
    function findWays(i, sum) {
        let cacheKey =  `${i}-${sum}`;
        if( cache.has(cacheKey) )
            return cache.get(cacheKey);
        
        if(  i === nums.length && sum === target )
            return 1
        
        if( i >= nums.length )
            return 0
        
        let res = 0;
        res += findWays(i+1, sum + nums[i]);
        res += findWays(i+1, sum - nums[i]);
        
        cache.set(cacheKey,res);
        return res;
    }
};